crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,1179 @@
+        assert_debug_snapshot!(non_stdlib_module_name_test_case("foo/bar/baz"), @r###"
+        ModuleName(
+            "foo.bar.baz",
+        )
+        "###);
+    }
+
+    #[test]
+    fn module_name_3_parts_pyi() {
+        assert_debug_snapshot!(non_stdlib_module_name_test_case("foo/bar/baz.pyi"), @r###"
+        ModuleName(
+            "foo.bar.baz",
+        )
+        "###);
+    }
+
+    #[test]
+    fn module_name_3_parts_py() {
+        assert_debug_snapshot!(non_stdlib_module_name_test_case("foo/bar/baz.py"), @r###"
+        ModuleName(
+            "foo.bar.baz",
+        )
+        "###);
+    }
+
+    #[test]
+    fn module_name_4_parts_dunder_init_pyi() {
+        assert_debug_snapshot!(non_stdlib_module_name_test_case("foo/bar/baz/__init__.pyi"), @r###"
+        ModuleName(
+            "foo.bar.baz",
+        )
+        "###);
+    }
+
+    #[test]
+    fn module_name_4_parts_dunder_init_py() {
+        assert_debug_snapshot!(non_stdlib_module_name_test_case("foo/bar/baz/__init__.py"), @r###"
+        ModuleName(
+            "foo.bar.baz",
+        )
+        "###);
+    }
+
+    #[test]
+    fn join_1() {
+        insta::with_settings!({filters => WINDOWS_PATH_FILTER.to_vec()}, {
+            assert_debug_snapshot!(
+                ModuleResolutionPathBuf::standard_library("foo").unwrap().join("bar"),
+                @r###"
+                ModuleResolutionPath::StandardLibrary(
+                    "foo/bar",
+                )
+                "###
+            );
+        });
+    }
+
+    #[test]
+    fn join_2() {
+        insta::with_settings!({filters => WINDOWS_PATH_FILTER.to_vec()}, {
+            assert_debug_snapshot!(
+                ModuleResolutionPathBuf::site_packages("foo").unwrap().join("bar.pyi"),
+                @r###"
+                ModuleResolutionPath::SitePackages(
+                    "foo/bar.pyi",
+                )
+                "###
+            );
+        });
+    }
+
+    #[test]
+    fn join_3() {
+        insta::with_settings!({filters => WINDOWS_PATH_FILTER.to_vec()}, {
+            assert_debug_snapshot!(
+                ModuleResolutionPathBuf::extra("foo").unwrap().join("bar.py"),
+                @r###"
+                ModuleResolutionPath::Extra(
+                    "foo/bar.py",
+                )
+                "###
+            );
+        });
+    }
+
+    #[test]
+    fn join_4() {
+        insta::with_settings!({filters => WINDOWS_PATH_FILTER.to_vec()}, {
+            assert_debug_snapshot!(
+                ModuleResolutionPathBuf::first_party("foo").unwrap().join("bar/baz/eggs/__init__.py"),
+                @r###"
+                ModuleResolutionPath::FirstParty(
+                    "foo/bar/baz/eggs/__init__.py",
+                )
+                "###
+            );
+        });
+    }
+
+    #[test]
+    #[cfg(debug_assertions)]
+    #[should_panic(expected = "Extension must be `pyi`; got `py`")]
+    fn stdlib_path_invalid_join_py() {
```

Comment:
I was worried that this would make `push()` and `join()` quite expensive. But your reasoning makes sense. If it's an invariant important enough to be tested (which I think it is), it should be asserted in release builds as well as debug builds. I've made the change.